### PR TITLE
Proj/patch for modern cmake

### PR DESCRIPTION
**Pull request by johnwparent**

### CODE REVIEW

**Comment by johnwparent** [comment]
cc. @alalazo Separated PROJ patch as discussed

**Review by alalazo — DISMISSED** [review]

**Review by adamjstewart — COMMENTED** [review]
Has anyone checked that this patch successfully applies to all versions? Or that it doesn't accidentally get applied to the autotools build?

**Comment by alalazo** [comment]
@adamjstewart Nope, I assumed that. I also didn't notice your review was required, sorry.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
> Has anyone checked that this patch successfully applies to all versions? 

The answer to this is no:
```console
> spack patch proj@5.0.1 build_system=cmake
==> Using cached archive: /Users/Adam/spack/var/spack/cache/_source-cache/archive/a7/a792f78897482ed2c4e2af4e8a1a02e294c64e32b591a635c5294cb9d49fdc8c.tar.gz
==> Using cached archive: /Users/Adam/spack/var/spack/cache/_source-cache/archive/3f/3ff6618a0acc9f0b9b4f6a62e7ff0f7bf538fb4f74de47ad04da1317408fcc15.tar.gz
==> Moving resource stage
	source: /var/folders/jv/cgkfvslj6nq1l7cw0c8c_8gm0000gn/T/Adam/spack-stage/resource-proj-datumgrid-4hst4kmzsncaht5zd2ishsbgjno7xqy3/spack-src/
	destination: /var/folders/jv/cgkfvslj6nq1l7cw0c8c_8gm0000gn/T/Adam/spack-stage/spack-stage-proj-5.0.1-4hst4kmzsncaht5zd2ishsbgjno7xqy3/spack-src/share/proj
==> Applied patch /Users/Adam/spack/var/spack/repos/builtin/packages/proj/proj.cmakelists.5.0.patch
No such line 404 in input file, ignoring
1 out of 1 hunks failed--saving rejects to 'src/lib_proj.cmake.rej'
==> Patch /Users/Adam/spack/var/spack/repos/builtin/packages/proj/proj-8.1-cmake-3.29-new-tiff-interface.patch failed.
==> Error: Command exited with status 1:
'/usr/bin/patch' '-s' '-p' '1' '-i' '/Users/Adam/spack/var/spack/repos/builtin/packages/proj/proj-8.1-cmake-3.29-new-tiff-interface.patch' '-d' '.'
```
We need to determine the minimum version of proj for which this patch actually works and set a lower bound.

> Or that it doesn't accidentally get applied to the autotools build?

From what I can tell, it won't accidentally get applied when autotools is used, but it may be worth guarding against just in case one of the dependencies depends on cmake.

**Comment by johnwparent** [comment]
> From what I can tell, it won't accidentally get applied when autotools is used, but it may be worth guarding against just in case one of the dependencies depends on cmake.

It can/will accidentally get added if you specifically request autotools for the `proj` build and `cmake` for another package in the dag. I initially thought the `^cmake` would solve for that but I was quite wrong.

File: var/spack/repos/builtin/packages/proj/package.py [hunk 1]
```diff
@@ -92,6 +92,11 @@ class Proj(CMakePackage, AutotoolsPackage):
 
     # https://proj.org/install.html#build-requirements
     with when("build_system=cmake"):
+        # CMake 3.19 refactored the FindTiff module interface, update older proj's
+        # to be compatible with this "new" interface
+        # patch replaces the TIFF_LIBRARY variable (no longer used) with TIFF_LIBRARIES
+        patch("proj-8.1-cmake-3.29-new-tiff-interface.patch", when="+tiff @:9.1.0 ^cmake@3.19:")
+        conflicts("cmake@3.19:", when="@:7")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/proj/package.py:99]: I can't remember if this needs to be `^cmake` or if `cmake` is also fine.

**Review by adamjstewart — APPROVED** [review]
I confirmed that proj 8.0.0 and 9.1.0 both patch correctly. I'll let @alalazo double check the syntax, but I think this is correct now. Thanks for the fix!

**Pull request merged**